### PR TITLE
Fix uprobe listing and attaching when target is in a separate mount namespace

### PR DESCRIPTION
**Pull request by dalehamel**
Fixes https://github.com/iovisor/bpftrace/issues/968, at least for uprobes.

I'll follow up with a separate issue and patches for USDT probes, which I think are being similarly mishandled when the target is in a separate mount namespace.

I wrote two new regression tests, which utilize a wrapper  (mountns_wrapper.c) that I made for the runtime tests which will create a new mount namespace, and privately bind mount the test program folder there. This simulates a docker container somewhat, as the path won't be visible to bpftrace, which is in a separate mount namespace. This is, I think, about the smallest, simplest wrapper I could use to reproduce the issue.

Both regression tests fail without my changes, and pass with them.

My changes are pretty simple:

* In the semantic analyzer, if a PID is specified, then the target path has /proc/PID/root prepended to it. This allows attaching to probes in where the binary being uprobed is in a different mount namespace. There don't seem to be any side effects to just prepending the /proc relative root path, as all of the other tests including runtime tests continue to pass.
* In the list code, there is already special handling for if a PID is specified. I just make the absolute path include the /proc/PID/root bits at the front, and it allows listing probes for a binary in a separate mount namespace.

@danobi @fbs @mmarchini please take a look when you have a chance

### CODE REVIEW

File: src/ast/semantic_analyser.cpp [hunk 1]
```diff
@@ -1240,7 +1240,14 @@ void SemanticAnalyser::visit(AttachPoint &ap)
     auto paths = resolve_binary_path(ap.target);
     if (paths.size() > 1)
       err_ << "path '" << ap.target << "' must refer to a unique binary but matched " << paths.size() << std::endl;
-    ap.target = paths.front();
+
+    if (bpftrace_.pid_ > 0) {
+      std::ostringstream abs;
+      abs << "/proc/" << bpftrace_.pid_ << "/root/" << paths.front();
```
Comment (fbs) [inline src/ast/semantic_analyser.cpp]: should we make a function that does this? Makes it easier to reuse/refactor when needed.

File: tests/testprogs/mountns_wrapper.c [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+    char* private_mount = "/tmp/unshare/";
+
+    if (unshare(CLONE_NEWNS | CLONE_FS) != 0)
+      errExit("Failed to unshare");
+
+    if (mount(NULL, "/", NULL, MS_REC | MS_PRIVATE, NULL) != 0)
+      errExit("Failed to make mount private");
+
+    if (mkdir(private_mount, S_IRWXG) != 0 && (errno != EEXIST))
```
Comment (fbs) [inline tests/testprogs/mountns_wrapper.c]: nit: personally prefer octal file modes (0770) over these weird S_IRWXG enums

File: tests/testprogs/mountns_wrapper.c [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+      errExit("Failed to make private mount dir");
+
+    char path[BUFFER_SIZE];
+    int size = readlink("/proc/self/exe", path, sizeof(path));
```
Comment (fbs) [inline tests/testprogs/mountns_wrapper.c]: `BUFFER_SIZE - 1` as your're writing to `path[size]` below.

File: tests/testprogs/mountns_wrapper.c [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+#define errExit(msg)    do { perror(msg); exit(EXIT_FAILURE); \
+                        } while (0)
+
+
+
+int main(int argc, char *argv[])
+{
+
+    char* private_mount = "/tmp/unshare/";
```
Comment (fbs) [inline tests/testprogs/mountns_wrapper.c]: I don't think this is unique enough to be run on peoples machines. `/tmp/bpftrace-unshare-test` is more likely to prevent conflicts and people will know where the file came from.

File: tests/testprogs/mountns_wrapper.c (hunk 3 above)
Comment (dalehamel) [reply tests/testprogs/mountns_wrapper.c]: thanks

File: tests/testprogs/mountns_wrapper.c (hunk 2 above)
Comment (dalehamel) [reply tests/testprogs/mountns_wrapper.c]: changed, better to be consistent. Thanks!

File: tests/testprogs/mountns_wrapper.c (hunk 4 above)
Comment (dalehamel) [reply tests/testprogs/mountns_wrapper.c]: I'll make sure it is unique too.

I don't think any of the other tests use /tmp as far as I know, if they do maybe a shared subfolder per test run could be a future standard?

File: src/ast/semantic_analyser.cpp (hunk 1 above)
Comment (dalehamel) [reply src/ast/semantic_analyser.cpp]: Thanks for the suggestion, I imagine this will probably be needed elsewhere at some point.

I put it in util.cpp

**Comment by dalehamel** [comment]
> Some quick comments

Thanks @fbs , sorry the C for mountns_wrapper was a little sloppy. Cleaned it up and ran through clang-format too, added some docs.

> Havent looked at the details of the unshare/mount calls yeet.

I added some comments, the tl;dr is that the unshare creates a new mount namespace, the namespace sets itself to private mount propagation, and then the testprog folder is mounted in a temporary directory, where it is then called from with execvp, replacing the mountns_wrapper process with the usual test program.

**Comment by fbs** [comment]
Minor but:

```
$ sudo ./mountns_wrapper ./uprobe_test &
$ sudo bpftrace -l -p $(pidof uprobe_test)  'u:*'

uprobe:/proc/13684/root//uprobe_test:__new_fopen
uprobe:/proc/13684/root//uprobe_test:__wmemcpy
```

The double `//` in here is a bit ugly.

Other than that it looks fine, although I'm no namespace expert.

> I'll follow up with a separate issue and patches for USDT probes, which I think are being similarly mishandled when the target is in a separate mount namespace.

Yse looks like it (on a static binary):

```
$ sudo bpftrace -l -p $(pidof uprobe_test)  | grep usdt
usdt:/uprobe_test:libc:memory_mallopt_arena_max
usdt:/uprobe_test:libc:memory_mallopt_arena_test
```

**Comment by dalehamel** [comment]
> The double // in here is a bit ugly.

I agree, I'll fix it.

**Comment by dalehamel** [comment]
It looks like the behavior of this code changed quite a lot, I'm working on a rebase

**Comment by dalehamel** [comment]
@fbs I addressed your comments on the double path separator

While I was working on this @acj landed 5005902 and 3f7a94a which created conflicts. @acj you might want to look this over / comment.

Because of 3f7a94a, I reworked this to add a new interface to `resolve_binary_path`, overloading it to allow optionally specifying a PID. If the PID is specified AND is in a different mount namespace, then
PATH is read based on the PID's own environment, rather than bpftrace's.

I couldn't figure out a good way to add a unit test for the new interface, as it requires spawning a child process and knowing its pid. I thought of doing this with `exec_system`, but as it stands the runtime tests cover this behavior.

I had to write a new helper function as well, to determine if the target id is in a different mount namespace.

The resulting refactor I think is actually more correct and a better iteration than the original, as it now only appends the /proc/PID/root bit if it knows for certain the target pid needs this.

File: src/list.cpp [hunk 5]
```diff
@@ -152,7 +152,7 @@ void list_probes(const BPFtrace &bpftrace, const std::string &search_input)
     if (bpftrace.pid_ > 0)
     {
       executable = get_pid_exe(bpftrace.pid_);
-      absolute_exe = executable;
+      absolute_exe = path_for_pid_mountns(bpftrace.pid_, executable);
```
Comment (dalehamel) [inline src/list.cpp]: this should maybe be changed to resolve_binary_path, wdyt?

Comment (dalehamel) [reply src/list.cpp]: i think i'll leave it as is for now, as this should always be safe to take the path relative to the pid's mount namespace, whether the mount namespace is the same as for bpftrace or note.

Comment (dalehamel) [reply src/list.cpp]: yeah nvm, resolve_binary_path returns a vector.

File: src/utils.cpp [hunk 6]
```diff
@@ -521,13 +566,69 @@ std::vector<std::string> resolve_binary_path(const std::string& cmd)
...
+likely that any references to local paths will not be valid, and that paths
+need to be made relative to the PID.
+
+If an invalid PID is specified or doesn't exist, it returns false.
+True is only returned if the namespace of the target process could be read and
+it doesn't match that of bpftrace.
+*/
+static bool pid_in_different_mountns(int pid) {
+
+  const size_t BUFSIZE = 4096;
```
Comment (dalehamel) [inline src/utils.cpp]: I agonized over this one. Realistically I don't expect this to ever exceed ever 32 bytes, as it is based on a formatted print of a known size, bounded by the the maximum PID size, which is currently 9 digits (4 million is max pid).

Likewise, reading the symlink should always return the strrep of an inode ID for the mount ns, again less than 32 bytes.

However, theoretically as these pare paths, i went with PATH_MAX. It's huge overkill, but probably still so fast to stack allocate that it doesn't matter.

File: src/utils.cpp [hunk 7]
```diff
@@ -77,6 +77,10 @@ static bool provider_cache_loaded = false;
 // Maps all providers of pid to vector of tracepoints on that provider
 static std::map<std::string, usdt_probe_list> usdt_provider_cache;
 
+static bool pid_in_different_mountns(int pid);
```
Comment (dalehamel) [inline src/utils.cpp]: I made this static for now because it hasn't been needed elsewhere, but this might be useful in other places in the code.

File: src/utils.cpp [hunk 8]
```diff
@@ -508,10 +511,53 @@ std::string exec_system(const char* cmd)
...
+std::vector<std::string> resolve_binary_path(const std::string& cmd, int pid)
+{
+  std::string env_paths = "";
+  std::ostringstream pid_environ_path;
+
+  if (pid_in_different_mountns(pid)) {
+    pid_environ_path << "/proc/" << pid << "/environ";
+    std::ifstream environ(pid_environ_path.str());
+
+    if(environ) {
```
[the remark this answers is not in the document]
Comment (dalehamel) [reply src/utils.cpp]: I ran the patch through clang-format, per #639 to try and avoid style nits like this

[the remark this answers is not in the document]
Comment (acj) [reply src/utils.cpp]: Sorry, I deleted this comment after seeing your second commit, but GH kept it for some reason. Please disregard.

File: src/utils.cpp (hunk 6 above)
Comment (dalehamel) [reply src/utils.cpp]: thanks for weighing in, i'll stick with 64.

File: src/utils.cpp [hunk 9]
```diff
@@ -521,13 +566,69 @@ std::vector<std::string> resolve_binary_path(const std::string& cmd)
...
+  else
+    return false;
+
+  sprintf(pid_path, "/proc/%d/ns/mnt", pid);
+  if (access(pid_path, F_OK) != -1) {
+    bytes = readlink(pid_path, pid_ns, BUFSIZE - 1);
+    if (bytes > 0 && bytes < BUFSIZE)
+      pid_ns[bytes] = '\0';
+    else
+      return false;
```
[the remark this answers is not in the document]
Comment (dalehamel) [reply src/utils.cpp]: Done, I rolled this up with the check for errno as well.

File: src/utils.cpp [hunk 10]
```diff
@@ -521,13 +566,69 @@ std::vector<std::string> resolve_binary_path(const std::string& cmd)
...
+  if (pid <= 0)
+    return false;
+
+  bytes = readlink("/proc/self/ns/mnt", my_ns, BUFSIZE - 1);
+  if (bytes > 0 && bytes < BUFSIZE)
+    my_ns[bytes] = '\0';
+  else
+    return false;
+
+  sprintf(pid_path, "/proc/%d/ns/mnt", pid);
```
[the remark this answers is not in the document]
Comment (dalehamel) [reply src/utils.cpp]: done

File: src/utils.cpp [hunk 11]
```diff
@@ -521,13 +566,69 @@ std::vector<std::string> resolve_binary_path(const std::string& cmd)
...
+  if (access(pid_path, F_OK) != -1) {
```
[the remark this answers is not in the document]
Comment (dalehamel) [reply src/utils.cpp]: good catch, I've dropped the `access` check and will just check errno when I check the edge cases on `bytes`

File: tests/testprogs/mountns_wrapper.c [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+
+  const char *private_mount = "/tmp/bpftrace-unshare-mountns-test";
+  char dpath[BUFFER_SIZE];
+  char exe[BUFFER_SIZE];
+
+  if (argc != 2)
+    errExit("Must specify test program as only argument.");
+
+  // Enter a new mount namespace
+  if (unshare(CLONE_NEWNS | CLONE_FS) != 0)
```
[the remark this answers is not in the document]
Comment (dalehamel) [reply tests/testprogs/mountns_wrapper.c]: good catch, saw than in `man 2 unshare` when I just checked:

> Specifying  this  flag  automatically implies CLONE_FS as well.

File: src/utils.cpp [hunk 13]
```diff
@@ -521,13 +567,70 @@ std::vector<std::string> resolve_binary_path(const std::string& cmd)
...
+std::string path_for_pid_mountns(int pid, const std::string &path) {
+  std::ostringstream pid_relative_path;
+  std::string root =
+      (path.length() >= 1 && path.at(0) == '/') ? "/root" : "/root/";
+  pid_relative_path << "/proc/" << pid << root << path;
+  return pid_relative_path.str();
+}
+
+/*
+Determines if the target process is in the same mount namespace as bpftrace.
```
[the remark this answers is not in the document]
Comment (dalehamel) [reply src/utils.cpp]: Good catch, I renamed the function (used to be `in_same_mountns`) and didn't change the comment.

File: src/utils.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (dalehamel) [reply src/utils.cpp]: All good, I'm looking forward to #639 being merged ;)

File: src/utils.cpp [hunk 14]
```diff
@@ -618,15 +616,15 @@ static bool pid_in_different_mountns(int pid)
     return false;
 
   sprintf(pid_path, "/proc/%d/ns/mnt", pid);
-  if(access( pid_path, F_OK ) != -1) {
+  if (access(pid_path, F_OK) != -1) {
```
[the remark this answers is not in the document]
Comment (dalehamel) [reply src/utils.cpp]: Code was removed

File: tests/testprogs/mountns_wrapper.c [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+    errExit("Failed to make private mount dir");
+
+  int idx = readlink("/proc/self/exe", dpath, sizeof(dpath) - 1);
+  dpath[idx] = '\0';
+
+  char *dname = dirname(dpath);
+  if (mount(dname, private_mount, NULL, MS_BIND, NULL) != 0)
+    errExit("Failed to set up private bind mount");
+
+  if ((strlen(private_mount) + strlen(argv[1])) > (BUFFER_SIZE - 1))
```
[the remark this answers is not in the document]
Comment (dalehamel) [reply tests/testprogs/mountns_wrapper.c]: Looking more closely, I realized I can just drop this check altogether and use `snprintf`, I also changed `BUFFER_SIZE` to `PATH_MAX`, as we are actually concatenating arbitrary paths here.

**Comment by dalehamel** [comment]
thanks for the review comments @fbs @acj I believe I've addressed them all

File: src/utils.cpp [hunk 16]
```diff
@@ -521,13 +566,70 @@ std::vector<std::string> resolve_binary_path(const std::string& cmd)
...
+    my_ns[bytes] = '\0';
+  else
+    return false;
+
+  snprintf(pid_path, BUFSIZE, "/proc/%d/ns/mnt", pid);
+  bytes = readlink(pid_path, pid_ns, BUFSIZE - 1);
+
+  if (errno == EACCES || errno == ENOENT || bytes > BUFSIZE) {
+    std::cerr << "Error reading mount ns from " << pid_path;
+    std::cerr << " for pid " << pid << std::endl;
```
Comment (fbs) [inline src/utils.cpp]: can you add the error here too? `strerror(errno)` iirc

Comment (dalehamel) [reply src/utils.cpp]: Done, I noticed this is failing on CI, I wonder if it will show a permission error on the next run...

**Comment by dalehamel** [comment]
@fbs I found the reason why it was failing. When I removed the call to `access` to check if the file exists, using `readlink` will give `ENOENT`, because the symlink is (by design) a dangling symlink.

I decided to ditch the approach of using `readlink` at all, and looked to see how bcc does this, here is a great example:

https://github.com/iovisor/bcc/blob/0aca2c71c9b9e6b3d3335b6df288ad01a09c3ed6/src/cc/libbpf.c#L881-L932

I adapted this code, dropping the GOTO, but preserving the behavior. This just reads and compares the inode numers directly. CI now passes.

As a bonus, using `perror` should help with reporting errors.

File: src/utils.cpp (hunk 16 above)
Comment (dalehamel) [reply src/utils.cpp]: I figured out the issue, comment below.

File: src/utils.cpp [hunk 17]
```diff
@@ -521,13 +575,92 @@ std::vector<std::string> resolve_binary_path(const std::string& cmd)
...
+    return false;
+
+  if ((size_t)snprintf(buf, sizeof(buf), "/proc/%d/ns/mnt", pid) >= sizeof(buf))
+    return false;
+
+  self_fd = open("/proc/self/ns/mnt", O_RDONLY);
+  if (self_fd < 0)
+  {
+    perror("open(/proc/self/ns/mnt)");
+    close(self_fd);
```
[the remark this answers is not in the document]
Comment (dalehamel) [reply src/utils.cpp]: I've gone with the goto suggestion, which I think handles this better.

File: src/utils.cpp [hunk 18]
```diff
@@ -521,13 +575,92 @@ std::vector<std::string> resolve_binary_path(const std::string& cmd)
...
+Determines if the target process is in a different mount namespace from
+bpftrace.
+
+If a process is in a different mount namespace (eg, container) it is very
+likely that any references to local paths will not be valid, and that paths
+need to be made relative to the PID.
+
+If an invalid PID is specified or doesn't exist, it returns false.
+True is only returned if the namespace of the target process could be read and
+it doesn't match that of bpftrace.
```
[the remark this answers is not in the document]
Comment (dalehamel) [reply src/utils.cpp]: I decided to add an exception, currently not catching it. I think that it will be more helpful than these errors.

Let me know what you think.

[the remark this answers is not in the document]
Comment (dalehamel) [reply src/utils.cpp]: I'll need to update the comment if you like the approach - reminder for myself.

File: src/utils.cpp [hunk 19]
```diff
@@ -521,13 +575,101 @@ std::vector<std::string> resolve_binary_path(const std::string& cmd)
...
+
+  if (fstat(self_fd, &self_stat))
+  {
+    errmsg = "fstat(self_fd)";
+    goto error;
+  }
+
+  if (fstat(target_fd, &target_stat))
+  {
+    errmsg = "fstat(target_fd)";
```
Comment (fbs) [inline src/utils.cpp]: Knowing why it failed can be useful

```suggestion
    errmsg << "fstat(target_fd): " << strerror(errno);
```

(need to make errmsg a stringstream for this)

File: src/utils.cpp [hunk 20]
```diff
@@ -508,10 +513,59 @@ std::string exec_system(const char* cmd)
...
+If a pid is specified, the binary path is taken relative to its own PATH if
+it is in a different mount namespace. Otherwise, the path is resolved relative
+to the local PATH env var for bpftrace's own mount namespace if it is set
+*/
+std::vector<std::string> resolve_binary_path(const std::string &cmd, int pid)
+{
+  std::string env_paths = "";
+  std::ostringstream pid_environ_path;
+
+  if (pid_in_different_mountns(pid))
```
Comment (fbs) [inline src/utils.cpp]:
```suggestion
  if (pid > 0 && pid_in_different_mountns(pid))
```

File: src/utils.cpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (fbs) [reply src/utils.cpp]: Looks ok to me, the same approach is used for WildcardException

File: src/utils.cpp (hunk 19 above)
Comment (dalehamel) [reply src/utils.cpp]: Added thanks for the suggestion

**Comment by dalehamel** [comment]
@fbs thanks for another round of reviews, I believe I've addressed your comments.

If it looks good I'd be happy to squash/rebase.

**Comment by dalehamel** [comment]
squashed+rebased
